iroh-net/src/netcheck/reportgen.rs

```diff
@@ -881,9 +887,23 @@ async fn check_captive_portal(dm: &RelayMap, preferred_relay: Option<RelayUrl>)
         }
     };
 
-    let client = reqwest::ClientBuilder::new()
-        .redirect(reqwest::redirect::Policy::none())
-        .build()?;
+    let mut builder = reqwest::ClientBuilder::new().redirect(reqwest::redirect::Policy::none());
+    if let Some(Host::Domain(domain)) = url.host() {
+        // Use our own resolver rather than getaddrinfo
+        //
+        // For some reason reqwest wants SocketAddr rather than IpAddr but then proceeds to
+        // ignore the port, extracting it from the URL instead.  We supply a dummy port.
+        //
+        // Ideally we would try to resolve **both** IPv4 and IPv6 rather than purely race
+        // them.  But our resolver doesn't support that yet.
+        let addrs: Vec<_> = dns_resolver
+            .lookup_ipv4_ipv6_staggered(domain, DNS_TIMEOUT, DNS_STAGGERING_MS)
+            .await?
+            .map(|ipaddr| SocketAddr::new(ipaddr, 80))
+            .collect();
+        builder = builder.resolve_to_addrs(domain, &addrs);
```

Comment:
Another option is to implement reqwests' `Resolver` trait for our resolver.  But we don't own the type so that's quite a bit of changes.  This keeps our changes targeted to this feature and is not really any worse.

Reply:
I really like the idea of being able to keep using `reqwest` and implementing our custom `Resolver`, perhaps allowing users to provide their own `Resolver` if they want to, perhaps under an unstable feature flag.

Is it feasible to lean in to reqwest even more heavily? It seems like [it supports HTTP upgrades](https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.upgrade) and thus we could use it in even more places.

If we manage to replace all usage of direct hyper with reqwest, we might be able to have one universal way of configuring custom DNS resolvers or HTTPS proxying.

Reply:
https://docs.rs/iroh-net/latest/iroh_net/endpoint/struct.Builder.html#method.dns_resolver is the current way to configure the DNS resolver for users.  That's what this is picking up.

I don't have a lot of opinion on reqwest use, I've seen people remove it in favour of more low-level control and to reduce the dependency count.  We also need to take into account https://github.com/n0-computer/iroh/issues/2901.

But if we can do everything with reqwest and there's benefits to it then I don't have anything against it.